core/drivers/crypto/caam/caam_key.c

```diff
@@ -0,0 +1,746 @@
+	}
+
+	/* Define blob operation direction */
+	if (out_key->is_blob)
+		opflag |= BLOB_ENCAPS;
+	else
+		opflag |= BLOB_DECAPS;
+
+	/* Build OP flags depending on the blob type */
+	switch (out_key->key_type) {
+	case CAAM_KEY_BLACK_ECB:
+		opflag |= PROT_BLOB_INFO(ECB);
+		break;
+	case CAAM_KEY_BLACK_CCM:
+		opflag |= PROT_BLOB_INFO(CCM);
+		break;
+	default:
+		return CAAM_BAD_PARAM;
+	}
+
+	/* Allocate the descriptor */
+	desc = caam_calloc_desc(BLOB_OP_DESC_ENTRIES);
+	if (!desc) {
+		KEY_TRACE("CAAM Context Descriptor Allocation error");
+		return CAAM_OUT_MEMORY;
+	}
+
+	status = caam_key_alloc(out_key);
+	if (status) {
+		KEY_TRACE("Key output allocation error");
+		goto err;
+	}
+
+	/* Define input and output buffer size */
+	if (out_key->is_blob) {
+		/*
+		 * For a blob operation, the input key size is the original key
+		 * size of the black key.
+		 * The output key size is the final blob size.
+		 */
+		input_buffer_size = in_key->sec_size;
+		output_buffer_size = out_key->buf.length;
+	} else {
+		/*
+		 * For an un-blob operation, the input key size is the original
+		 * key size of the black key.
+		 * The output key size is the key security size.
+		 */
+		input_buffer_size = in_key->buf.length;
+		output_buffer_size = out_key->sec_size;
+	}
+
+	/* Create the blob encapsulation/decapsulation descriptor */
+	caam_desc_init(desc);
+	caam_desc_add_word(desc, DESC_HEADER(0));
+
+	/* Load the key modifier */
+	caam_desc_add_word(desc,
+			   LD_NOIMM(CLASS_2, REG_KEY, KEY_BLOB_MODIFIER_SIZE));
+	caam_desc_add_ptr(desc, virt_to_phys((void *)key_blob_modifier));
+
+	/* Define the Input data sequence */
+	caam_desc_add_word(desc, SEQ_IN_PTR(input_buffer_size));
+	caam_desc_add_ptr(desc, in_key->buf.paddr);
+
+	/* Define the Output data sequence */
+	caam_desc_add_word(desc, SEQ_OUT_PTR(output_buffer_size));
+	caam_desc_add_ptr(desc, out_key->buf.paddr);
+	caam_desc_add_word(desc, opflag);
+
+	KEY_DUMPDESC(desc);
+
+	caam_key_cache_op(TEE_CACHECLEAN, in_key);
+	caam_key_cache_op(TEE_CACHECLEAN, out_key);
+
+	jobctx.desc = desc;
+	status = caam_jr_enqueue(&jobctx, NULL);
+
+	if (status == CAAM_NO_ERROR) {
+		KEY_TRACE("CAAM Blob %scapsulation Done",
+			  out_key->is_blob ? "En" : "De");
+
+		caam_key_cache_op(TEE_CACHEINVALIDATE, out_key);
+		caam_key_dump("Blob output key", out_key);
+
+		goto out;
+	} else {
+		KEY_TRACE("CAAM Blob Status 0x%08" PRIx32 "", jobctx.status);
+	}
+
+err:
+	caam_key_free(out_key);
+out:
+	caam_free_desc(&desc);
+	return status;
+}
+
+enum caam_status caam_key_deserialize_from_bin(uint8_t *data, size_t size,
+					       struct caamkey *key,
+					       size_t sec_size)
+{
+	enum caam_status status = CAAM_FAILURE;
+	struct caamkey blob = { };
+
+	assert(data && size && key);
+
+	KEY_TRACE("Deserialization binary buffer");
+	KEY_DUMPBUF("Deserialize key buffer input", data, size);
+
+	/*
+	 * If a security key size is given, use it. Otherwise, rely on
+	 * the buffer size.
+	 */
+	if (sec_size == 0)
```

Comment:
This is needed because some keys will have leading zeros we must preserve. There is a comment few lines below. Let me know if you want me to move it

Reply:
I'm gonna check it

Reply:
We cannot rely on the bignum size as it could be smaller than the security key size. In this case, we need to specify the security key to properly do the bignum to bin translation and eventually pad the result. It was already done before :
https://github.com/OP-TEE/optee_os/blob/8e9d8acca4f8141b72d2d32846317954124c17ec/core/drivers/crypto/caam/acipher/caam_ecc.c#L108-L110

Reply:
Please leave an explanation like this in the comment above for future reference.